### PR TITLE
feat: get non-missed slot strategy

### PR DESCRIPTION
**Pull request by vgorkavenko**

### CODE REVIEW

**Comment by F4ever** [comment]
+8,475 −4,177

File: src/utils/slot.py [hunk 1]
```diff
@@ -17,37 +18,108 @@ class NoSlotsAvailable(Exception):
 def get_first_non_missed_slot(
     cc: ConsensusClient,
     ref_slot: SlotNumber,
+    last_finalized_slot_number: SlotNumber,
```
Comment (F4ever) [inline src/utils/slot.py:21]: What if after ref_slot there is no finalized slots, only unfinalized?
And what to put to last_finalized_slot if we don't know it?

File: src/providers/consensus/client.py [hunk 2]
```diff
@@ -38,6 +40,20 @@ def get_block_root(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockRo
         data, _ = self._get(self.API_GET_BLOCK_ROOT.format(state_id))
         return BlockRootResponse(**data)
 
+    def get_block_header(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockHeaderFullResponse:
+        """
+        Spec: https://ethereum.github.io/beacon-APIs/#/Beacon/getBlockRoot
```
Comment (F4ever) [inline src/providers/consensus/client.py:45]: Wron url

File: src/providers/consensus/client.py [hunk 3]
```diff
@@ -38,6 +40,20 @@ def get_block_root(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockRo
...
+
+        No cache because this method is using to get finalized and head block, and they could not be cached by args.
+        """
+        data, rest_response = self._get(self.API_GET_BLOCK_HEADER.format(state_id))
+        resp = BlockHeaderFullResponse(data=BlockHeaderResponseData(**data), **rest_response)
+        if hasattr(resp, 'finalized') and not resp.finalized:
+            raise Exception(f'Slot [{state_id}] is not finalized')
```
Comment (F4ever) [inline src/providers/consensus/client.py:52]: Too wide exceptions

File: src/providers/consensus/client.py [hunk 4]
```diff
@@ -38,6 +40,20 @@ def get_block_root(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockRo
...
 
+    def get_block_header(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockHeaderFullResponse:
+        """
+        Spec: https://ethereum.github.io/beacon-APIs/#/Beacon/getBlockRoot
+
+        No cache because this method is using to get finalized and head block, and they could not be cached by args.
+        """
+        data, rest_response = self._get(self.API_GET_BLOCK_HEADER.format(state_id))
+        resp = BlockHeaderFullResponse(data=BlockHeaderResponseData(**data), **rest_response)
+        if hasattr(resp, 'finalized') and not resp.finalized:
```
Comment (F4ever) [inline src/providers/consensus/client.py:51]: Why here is such check?

File: src/providers/consensus/client.py [hunk 5]
```diff
@@ -38,6 +40,20 @@ def get_block_root(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockRo
         data, _ = self._get(self.API_GET_BLOCK_ROOT.format(state_id))
         return BlockRootResponse(**data)
 
+    def get_block_header(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockHeaderFullResponse:
+        """
+        Spec: https://ethereum.github.io/beacon-APIs/#/Beacon/getBlockRoot
+
+        No cache because this method is using to get finalized and head block, and they could not be cached by args.
```
Comment (F4ever) [inline src/providers/consensus/client.py:47]: As I see we don't use this method with finalized and head params, so can use cache here

File: src/providers/consensus/client.py [hunk 6]
```diff
@@ -38,6 +40,20 @@ def get_block_root(self, state_id: Union[str, SlotNumber, BlockRoot]) -> BlockRo
...
+        """
+        data, rest_response = self._get(self.API_GET_BLOCK_HEADER.format(state_id))
+        resp = BlockHeaderFullResponse(data=BlockHeaderResponseData(**data), **rest_response)
+        if hasattr(resp, 'finalized') and not resp.finalized:
+            raise Exception(f'Slot [{state_id}] is not finalized')
+        if not resp.data.canonical:
```
Comment (F4ever) [inline src/providers/consensus/client.py:53]: And this check?

File: src/utils/slot.py (hunk 1 above)
Comment (vgorkavenko) [reply src/utils/slot.py:21]: A case when between `ref_slot` and `last_finalized_slot` no one finalized slot is determined as an exception.
We always know `last_finalized_slot` in each oracle cycle. It's equal `BaseModule._previous_finalized_slot_number`

File: src/providers/consensus/client.py (hunk 4 above)
Comment (vgorkavenko) [reply src/providers/consensus/client.py:51]: Perhaps I should move it. I think we’ll want to get some unfinalized slots somewhere

File: src/providers/consensus/client.py (hunk 6 above)
Comment (vgorkavenko) [reply src/providers/consensus/client.py:53]: Perhaps I should move it

**Comment by vgorkavenko** [comment]
> +8,475 −4,177 =)
> Pls, remove huge fixtures

fixed. left only needed changes in mocked responses

File: src/providers/consensus/client.py (hunk 6 above)
Comment (vgorkavenko) [reply src/providers/consensus/client.py]: fixed

File: src/providers/consensus/client.py (hunk 5 above)
Comment (vgorkavenko) [reply src/providers/consensus/client.py]: fixed

File: src/providers/consensus/client.py (hunk 4 above)
Comment (vgorkavenko) [reply src/providers/consensus/client.py]: fixed

File: src/providers/consensus/client.py (hunk 3 above)
Comment (vgorkavenko) [reply src/providers/consensus/client.py]: fixed

File: src/providers/consensus/client.py (hunk 2 above)
Comment (vgorkavenko) [reply src/providers/consensus/client.py]: fixed

**Review by F4ever — APPROVED** [review]

**Pull request merged**